aten/src/ATen/cuda/tunable/Tunable.cpp

```diff
@@ -0,0 +1,515 @@
+  if (env != nullptr) {
+    return atoi(env);
+  }
+  return max_tuning_duration_ms_;
+}
+
+void TuningContext::SetMaxTuningIterations(int max_iter) {
+  max_tuning_iterations_ = max_iter;
+}
+
+int TuningContext::GetMaxTuningIterations() const {
+  static const char *env = std::getenv("PYTORCH_TUNABLEOP_MAX_TUNING_ITERATIONS");
+  if (env != nullptr) {
+    return atoi(env);
+  }
+  return max_tuning_iterations_;
+}
+
+void TuningContext::SetMaxWarmupDurationMs(int max_duration_ms) {
+  max_warmup_duration_ms_ = max_duration_ms;
+}
+
+int TuningContext::GetMaxWarmupDurationMs() const {
+  static const char *env = std::getenv("PYTORCH_TUNABLEOP_MAX_WARMUP_DURATION_MS");
+  if (env != nullptr) {
+    return atoi(env);
+  }
+  return max_warmup_duration_ms_;
+}
+
+void TuningContext::SetMaxWarmupIterations(int max_iter) {
+  max_warmup_iterations_ = max_iter;
+}
+
+int TuningContext::GetMaxWarmupIterations() const {
+  static const char *env = std::getenv("PYTORCH_TUNABLEOP_MAX_WARMUP_ITERATIONS");
+  if (env != nullptr) {
+    return atoi(env);
+  }
+  return max_warmup_iterations_;
+}
+
+void TuningContext::SetBufferRotationCount(int count) {
+  buffer_rotation_count_ = count;
+}
+
+int TuningContext::GetBufferRotationCount() const {
+  static const char *env = std::getenv("PYTORCH_TUNABLEOP_BUFFER_ROTATION_COUNT");
+  if (env != nullptr) {
+    return atoi(env);
+  }
+  return buffer_rotation_count_;
+}
+
+void TuningContext::EnableTunableOpAndTuning() {
+  EnableTunableOp();
+  EnableTuning();
+}
+
+void TuningContext::DisableTunableOpAndTuning() {
+  DisableTunableOp();
+  DisableTuning();
+}
+
+TuningResultsManager& TuningContext::GetTuningResultsManager() {
+  c10::call_once(manager_init_once_, [this]() {
+    if (!GetFilename().empty()) {
+      ReadFile(GetFilename());
+    }
+  });
+  return manager_;
+}
+
+TuningResultsValidator& TuningContext::GetTuningResultsValidator() {
+  return validator_;
+}
+
+TuningResults TuningContext::GetTuningResults() {
+  TuningResults tr;
+  tr.validators = GetTuningResultsValidator().GetAllValidators();
+  tr.results = GetTuningResultsManager().Dump();
+  return tr;
+}
+
+TuningStatus TuningContext::LoadTuningResults(const TuningResults& tr) {
+  TORCH_CHECK(GetTuningResultsValidator().ValidateAll(tr.validators));
+  GetTuningResultsManager().Load(tr.results);
+  return OK;
+}
+
+void TuningContext::SetFilename(const std::string& filename) {
+  filename_ = filename;
+}
+
+std::string TuningContext::GetFilename() const {
+  static const char *env = std::getenv("PYTORCH_TUNABLEOP_FILENAME");
+  if (env != nullptr) {
+    return env;
+  }
+  return filename_;
+}
+
+void TuningContext::ReadFile(const std::string& filename) {
+  TUNABLE_LOG("reading tuning results from ", filename);
+  ResultsMap results;
+  std::unordered_map<std::string, std::string> validators;
+  std::string line;
+  std::ifstream file(filename);
+  while (std::getline(file, line)) {
```

Comment:
i wonder if we want to make this part a bit more robust. I remember if I put a random file, the program will seg fault rather than just ignore and re-profile (can put the crash stack if needed).

Reply:
Stack crash may have been due to trying to query the device ordinal during static destruction of TunableOp.  In any case, I have fixed both the device ordinal query to occur much earlier, and I have also made the file I/O more friendly by checking the file handle is good prior to reading, and I've also added a check that the file can be written and issue a warning at the beginning rather than at the end.  Neither failure to read or write is fatal.

2b36d8fc8c542113cd707e60e474bef04bf16a34 makes the I/O more friendly.